### PR TITLE
Stable diffusion pipeline

### PR DESCRIPTION
**Pull request by patil-suraj**
This PR adds `StableDiffusionPipeline`

### CODE REVIEW

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me!

**Comment by pcuenca** [comment]
Looks good to me! One minor question: since this is only for `StableDiffusion`, would it make sense to assume the scheduler has no `eta` and remove that argument and the `inspect`?

**Comment by patrickvonplaten** [comment]
> Looks good to me! One minor question: since this is only for `StableDiffusion`, would it make sense to assume the scheduler has no `eta` and remove that argument and the `inspect`?
+1 

Good remark - think this makes sense! It'll make the code cleaner and we could always add it later to the party

**Comment by patil-suraj** [comment]
Good point @pcuenca ! I agree with this, but since the original codebase allows to play with both DDIM and PNDM, think it would important to have the `eta` there. I left a more detailed comment in the code about when `eta` will be used.

**Comment by patil-suraj** [comment]
@patrickvonplaten @anton-l 
Could you take one more look, updated the pipeline a bit to make it more cleaner:

1. Simplify classifier free guidance.
2. Add more comments
3. add type hints and change ordering of parameters.
4. remove unused parameters.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        do_classifier_free_guidance = guidance_scale > 1.0
+        # get unconditional embeddings for classifier free guidance
+        if do_classifier_free_guidance:
+            uncond_input = self.tokenizer([""] * batch_size, padding="max_length", max_length=77, return_tensors="pt")
+            uncond_embeddings = self.text_encoder(uncond_input.input_ids.to(torch_device))[0]
+
+            # For classifier free guidance, we need to do two forward passes.
+            # Here we concatenate the unconditional and text embeddings into a single batch
+            # to avoid doing two forward passes
+            text_embeddings = torch.cat((uncond_embeddings, text_embeddings), dim=0)
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:61]: doing the concatenation here rather than in the loop, to avoid doing it multiple times.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        extra_kwrags = {}
+        if accepts_eta:
+            extra_kwrags["eta"] = eta
+
+        self.scheduler.set_timesteps(num_inference_steps)
+
+        for t in tqdm(self.scheduler.timesteps):
+            # expand the latents if we are doing classifier free guidance
+            if do_classifier_free_guidance:
+                latents = torch.cat((latents, latents), dim=0)
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:82]: since `text_embeddings` are already expanded, we only need to expand `latents` here.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        latents = torch.randn(
+            (batch_size, self.unet.in_channels, self.unet.sample_size, self.unet.sample_size),
+            generator=generator,
+        )
+        latents = latents.to(torch_device)
+
+        # prepare extra kwargs for the scheduler step, since not all schedulers have the same signature
+        # eta (η) is only used with the DDIMScheduler, it will be ignored for other schedulers.
+        accepts_eta = "eta" in set(inspect.signature(self.scheduler.step).parameters.keys())
+        extra_kwrags = {}
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:73]: Typo here (and in subsequent lines) `kwrags` -> `kwargs`

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        self,
+        prompt: Union[str, List[str]],
+        num_inference_steps: Optional[int] = 50,
+        guidance_scale: Optional[float] = 1.0,
+        eta: Optional[float] = 0.0,
+        generator: Optional[torch.Generator] = None,
+        torch_device: Optional[Union[str, torch.device]] = None,
+        output_type: Optional[str] = "pil",
+    ):
+        # eta corresponds to η in paper and should be between [0, 1]
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:38]: maybe remove this comment and move the interval to line [71](https://github.com/huggingface/diffusers/pull/168/files#diff-ab952f41078da66b9fcbbd913b419f8c334badceefac03a5f7edcd6dd986a8efR71)

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 5]
```diff
@@ -35,47 +35,49 @@ def __call__(
         self.vae.to(torch_device)
         self.text_encoder.to(torch_device)
 
+        # get prompt text embeddings
+        text_input = self.tokenizer(prompt, padding="max_length", max_length=77, return_tensors="pt")
+        text_embeddings = self.text_encoder(text_input.input_ids.to(torch_device))[0]
+
+        do_classifier_free_guidance = guidance_scale > 1.0
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:52]: We are now ignoring values of `guidance_scale` < 1, I suppose that makes sense.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 6]
```diff
@@ -1,28 +1,39 @@
...
 
 class StableDiffusionPipeline(DiffusionPipeline):
-    def __init__(self, vae, text_encoder, tokenizer, unet, scheduler):
+    def __init__(
+        self,
+        vae: AutoencoderKL,
+        text_encoder: CLIPTextModel,
+        tokenizer: CLIPTokenizer,
+        unet: UNet2DConditionModel,
+        scheduler: Union[DDIMScheduler, PNDMScheduler],
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:21]: 👍

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:52]: If I understand correctly in the original paper the guidance scale `w` (in the paper called `"guidance strength"`) was chosen between [0, inf) - see: equation 6 here: https://arxiv.org/pdf/2207.12598.pdf) whereas it seems like CompVis used an adapted formula where `scale = 1 + w` so that scale = 1 => w = 0 => no classifier free guidance. 

In the Imagen paper they also use the CompVis notation where `w = 1 => no classifier free guidance`. Think we should be quite specific here what exactly we mean by `guidance_scale` then to avoid confusion.

IMO we should either 

```suggestion
         # here `guidance_scale` is defined analog to the guidance weight `w` of equation (2)
         # of the Imagen paper: https://arxiv.org/pdf/2205.11487.pdf . `guidance_scale = 1`
         # corresponds to doing no classifier free guidance.
        do_classifier_free_guidance = guidance_scale > 1.0
```

Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:52]: In the CompVis paper they aren't very clear about it (no formula is shown: https://arxiv.org/pdf/2112.10752.pdf)

**Review by patrickvonplaten — APPROVED** [review]

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        accepts_eta = "eta" in set(inspect.signature(self.scheduler.step).parameters.keys())
+        extra_kwargs = {}
+        if accepts_eta:
+            extra_kwargs["eta"] = eta
+
+        self.scheduler.set_timesteps(num_inference_steps)
+
+        for t in tqdm(self.scheduler.timesteps):
+            # expand the latents if we are doing classifier free guidance
+            latent_model_input = torch.cat([latents] * 2) if do_classifier_free_guidance else latents
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:93]: @patil-suraj there was a bug with the naming here -> corrected it

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        guidance_scale: Optional[float] = 1.0,
+        eta: Optional[float] = 0.0,
+        generator: Optional[torch.Generator] = None,
+        torch_device: Optional[Union[str, torch.device]] = None,
+        output_type: Optional[str] = "pil",
+    ):
+        if torch_device is None:
+            torch_device = "cuda" if torch.cuda.is_available() else "cpu"
+
+        if isinstance(prompt, str):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:41]: @patil-suraj -> made sure a string can be passed as input

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            batch_size = len(prompt)
+        else:
+            raise ValueError(f"`prompt` has to be of type `str` or `list` but is {type(prompt)}")
+
+        self.unet.to(torch_device)
+        self.vae.to(torch_device)
+        self.text_encoder.to(torch_device)
+
+        # get prompt text embeddings
+        text_input = self.tokenizer(prompt, padding=True, truncation=True, return_tensors="pt")
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:53]: @patil-suraj removed the weird 77 max length here

**Review by patrickvonplaten — APPROVED** [review]
Merging now! Tested it and it works fine

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py]: that gave a problem with naming but should be fixed now

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 5 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:59]: Thanks a lot for tracking this down, @patrickvonplaten! I think the comment is very helpful and prevents uncertainty.
